### PR TITLE
Add kubeadm init unit test

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
as per this [comment](https://github.com/kubernetes/kubernetes/pull/72442#discussion_r244722092) the PR adds unit tests for kubeadm init (and more specifically for `newInitData` method that implements all the command options validation logic)

**Special notes for your reviewer**:
the PR includes to minor cleanup/fix in init.go file (L199, L296)

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig cluster-lifecycle
/priority backlog
/assign @yagonobre 
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
thanks @fabriziopandini 
/lgtm

**Comment by neolit123** [comment]
/hold
adding a quick hold for more comments if needed.

File: cmd/kubeadm/app/cmd/init_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+func TestNewInitData(t *testing.T) {
+	// create temp directory
+	tmpDir, err := ioutil.TempDir("", "kubeadm-init-test")
+	if err != nil {
+		t.Errorf("Unable to create temporary directory: %v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// create config file
```
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/init_test.go]: Optionally you can use
```golang
manifestPath := filepath.Join(tmpdir, "test-config-file")
ioutil.WriteFile(manifestPath, []byte(testInitConfig)), 0644)
```

File: cmd/kubeadm/app/cmd/init_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+			name: "fails if invalid preflight checks are provided",
+			flags: map[string]string{
+				options.IgnorePreflightErrors: "all,something-else",
+			},
+			expectError: true,
+		},
+
+		// Init data passed using config file
+		{
+			name: "Pass with config from file",
```
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/init_test.go]: start with lowercase

**Review by yagonobre — APPROVED** [review]
Thanks @fabriziopandini, just non blocking comments
/lgtm

**Comment by fabriziopandini** [comment]
@neolit123 is it possible to remove hold or there still is something to do that I'm missing ?

**Comment by neolit123** [comment]
@fabriziopandini feel free to remove hold.
i've added it so that you can control the merge.

File: cmd/kubeadm/app/cmd/init_test.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			},
+			expectError: true,
+		},
+		{
+			name: "fail if deprecetes feature gates are set",
+			flags: map[string]string{
+				options.FeatureGatesString: fmt.Sprintf("%s=true", features.CoreDNS),
+			},
+			expectError: true,
+		},
```
Comment (rosti) [inline cmd/kubeadm/app/cmd/init_test.go]: We should have a TODO to add a positive test here, when actual non-deprecated feature gates are added to kubeadm.

File: cmd/kubeadm/app/cmd/init_test.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		{
+			name: "fails if invalid preflight checks are provided",
+			flags: map[string]string{
+				options.IgnorePreflightErrors: "all,something-else",
+			},
+			expectError: true,
+		},
```
Comment (rosti) [inline cmd/kubeadm/app/cmd/init_test.go]: We should add a positive test here, that validates, that valid values for the `IgnorePreflightErrors` are observed. Can be just a single value, the idea is to test if the flag gets propagated and not dismissed.

**Review by rosti — COMMENTED** [review]
We may need to add a few more positive test cases. We should probably test `--skip-phases` too.
At some point in the future, we may also merge tests here with the ones in this file:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/test/cmd/init_test.go

**Comment by fabriziopandini** [comment]
@rosti 
this train is gone, but follow up PRs are welcome, but IMO we should wait for join refactor to merge before so test can be duplicated (if it makes sense).

I'm also ok in removing test in `test/cmd/*` because they are typically slow than unit test introduced in this PR (assuming they are functionally equivalent)

**Pull request merged**